### ISSUE TITLE
Lazy loading of data for model

### ISSUE THREAD

**Issue by thequackdaddy** [statsmodels/statsmodels#3609]
Hello,

Thanks to an excellent suggestion by @njsmith over at the patsy page, you can have larger, on-disk data held for later evaluation using bcolz. Take this toy script:

```python
import bcolz
import os
import pandas as pd
import statsmodels.formula.api as smf
import statsmodels.api as sm
import odo

cpunish = sm.datasets.cpunish.load_pandas().data
cpunish = odo.odo(cpunish, 'c:/cpunish.bcolz/')

class LazyBcolz(dict):
    def __init__(self, file):
        if not os.path.exists(file):
            raise IOError('path \'{}\' not found'.format(file))
        self.file = file
    
    def __missing__(self, column):
        path = os.path.join(self.file, column)
        if os.path.exists(path):
            return pd.Series(bcolz.open(path)[:], name=column)
        else:
            raise KeyError('{} not in ctable'.format(column))

cpunish_lazy = LazyBcolz('c:/cpunish.bcolz')

model = smf.glm('EXECUTIONS ~ INCOME + PERPOVERTY + PERBLACK + SOUTH',
                cpunish_lazy)

res = model.fit()
```
Defining the `LazyBcolz` class allows me to store my data on disk and lazily load only the columns necessary to build the design matrix. This helps keep memory consumption way down. 

So far, this script works perfectly. 

Sadly, however, I can't get the `predict` method to work appropriately. 
```
res.predict(cpunish_lazy)
```
It appears that `predict` *used* to work before [this commit](https://github.com/statsmodels/statsmodels/commit/43e47cb244a34362612c3775694f440ba70c0717) to handle `pd.Series` inadvertently changed this. 

So I have a few options here...

* Is this type of approach to lazily load data something that could/should supported? Playing with it, it seems really handy to me. 
* Is this something that should be considered for incorporation in the #3203. I don't *think* this functionality is currently supported, although it could be. 
* Alternatively, is there a to write a class which is an extension of dataframes (so `isinstance(obj, pd.DataFrame)== True`) but have it still load data lazily? 

-> Lazy work-around, use patsy `res.predict(dmatrix(model.data.design_info, cpunish_lazy), transform=False)`, but that feels like a lot of typing. 

At a minimum, I'd wonder if writing some unit tests of the model set-up and fit would be welcome to keep this handy functionality.

**Comment by josef-pkt** [comment]
what's `type(model.data.exog_orig)` if I remember the attribute correctly?
or what type does `dmatrix(model.data.design_info, cpunish_lazy)` return?

My guess without checking anything is that we need also an `index` for pandas look-alikes. My main target in this area would be to do define some requirements for necessary information without full type/isinstance checking. 
e.g. what specific part makes `predict` fail in this case? Why don't we go through patsy dmatrix in this case anyway?

Do `res.resid` `res.fittedvalues` get a correct `index`?

AFAICS, this is just preprocessing, once `exog` or the `dmatrix` array_like is created it is all in-memory and shouldn't affect anything.

**Comment by thequackdaddy** [comment]
> what's type(model.data.exog_orig) if I remember the attribute correctly?

If you mean `model.data.frame`, it gives `__main__.LazyBcolz`

> what specific part makes predict fail in this case? Why don't we go through patsy dmatrix in this case anyway?

It fails [here](https://github.com/statsmodels/statsmodels/blob/master/statsmodels/base/model.py#L804)

What happens is that a few lines up we do, `exog = pd.DataFrame(exog)`... this returns an empty dataframe. Then, when `dmatrix` is called, all the variables are missing so `patsy` complains. 

> Do res.resid res.fittedvalues get a correct index?

Yes.

(Note that these appear to call `predict` method from the `model` object, while `res.predict` is on the `result` object. The `model` object seems like it handles this well. the `result` object... not so well.)

**Comment by josef-pkt** [comment]
To the note: the `model.predict` assumes in general that we have a numpy ndarray, Otherwise, the result might work or not depending on how exog quacks.

The rewrite of result.predict turned out to be quite messy, because there are many different cases to handle. 
I think, instead of converting a user provided exog to pandas, `exog = pd.DataFrame(exog)` we could hand off the original exog to patsy dmatrix and let patsy check that it is compatible with the original specification of the formula and data.
The main problem that causes a bit of convoluted code here is that we need to work around patsy's dropping of nan observations. We want the return of predict with the original length and index including nan rows/elements.
My guess is still that all we need is a data structure with an index. If exog has an index, then we don't need to convert to pandas to get an index.

**Comment by TomAugspurger** [comment]
@thequackdaddy do you have any experience with dask? I suspect (just a hunch) that it will require less effort to get statsmodels working with dask arrays / dataframes, though it does seem like you're pretty close already. If only I hadn't broken it with that commit :/

I have a start [here](https://github.com/statsmodels/statsmodels/commit/11e5018f6de82daed524337f5f70d9b4f667be65) for accepting dask-arrays. If you have a need for it, I can pick that work up again.

**Comment by josef-pkt** [comment]
@TomAugspurger  AFAIU, these are two different although related use cases. Here, we just want to support out of memory pre-processing, where the core is in-memory. Your dask work, AFAIU, prepares for internal usage of out of memory computation. 

For the former we mainly need a more flexible interface but don't need to change the internals, e.g. patsy could support creating the design matrix directly from a database connection, but the actual model is in-memory.

**Comment by thequackdaddy** [comment]
@TomAugspurger Thanks for the link. I'll look at this. I think it was pretty much (accidentally) "feature complete" so I'm not sure if I want to go about the dask route yet. A few months ago (before I asked for some additional RAM on a virtual machine I run) this was an issue. But now, I have pretty much all the RAM I need.

**Comment by thequackdaddy** [comment]
Closed by #3611
